File: packages/sdk/src/testing/runner.ts
/**
 * Policy unit testing framework.
 *
 * Reads YAML fixture files and evaluates each test case against the policy
 * using the deterministic replay engine (no LLM, no network).
 *
 * @module testing/runner
 */

import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { resolve, join } from 'node:path';
import { parse as parseYaml } from 'yaml';
import {
	buildReplaySnapshot,
	decideReplayCall,
	loadPolicySnapshot,
} from '../cli/replay-engine.js';
import type { ASTNode } from '../compiler/index.js';
import type { ReplayCall } from '../cli/replay-engine.js';
import type {
	VetoTestCase,
	VetoTestResult,
	VetoTestRunResult,
	VetoTestSuite,
} from './types.js';
import { colors } from '../cli/colors.js';

// ---------------------------------------------------------------------------
// Normalization
// ---------------------------------------------------------------------------

/**
 * The replay engine action `block` surfaces as decision `deny`.
 * `warn` and `log` actions surface as `allow` (non-blocking).
 * Normalize the expected value from a fixture to what the engine returns.
 */
function normalizeExpectedDecision(
	expected: VetoTestCase['expect']['decision'],
): string {
	switch (expected) {
		case 'block':
		case 'deny':
			return 'deny';
		case 'warn':
		case 'log':
			return 'allow';
		default:
			return expected;
	}
}

// ---------------------------------------------------------------------------
// Fixture loading
// ---------------------------------------------------------------------------

function collectYamlFiles(dir: string): string[] {
	const results: string[] = [];
	if (!existsSync(dir)) return results;

	const entries = readdirSync(dir);
	for (const entry of entries) {
		const full = join(dir, entry);
		const stat = statSync(full);
		if (stat.isDirectory()) {
			results.push(...collectYamlFiles(full));
		} else if (entry.endsWith('.yaml') || entry.endsWith('.yml')) {
			results.push(full);
		}
	}
	return results;
}

function loadSuites(fixturesPath: string): { suites: VetoTestSuite[]; errors: string[] } {
	const suites: VetoTestSuite[] = [];
	const errors: string[] = [];

	let files: string[];
	const resolved = resolve(fixturesPath);

	if (!existsSync(resolved)) {
		errors.push(`Fixture path not found: ${resolved}`);
		return { suites, errors };
	}

	const stat = statSync(resolved);
	if (stat.isFile()) {
		files = [resolved];
	} else {
		files = collectYamlFiles(resolved);
		if (files.length === 0) {
			errors.push(`No YAML fixture files found in: ${resolved}`);
		}
	}

	for (const file of files) {
		let content: string;
		try {
			content = readFileSync(file, 'utf-8');
		} catch (err) {
			errors.push(`Failed to read fixture file ${file}: ${err instanceof Error ? err.message : String(err)}`);
			continue;
		}

		let parsed: unknown;
		try {
			parsed = parseYaml(content);
		} catch (err) {
			const yamlErr = err as { mark?: { line?: number; column?: number }; message?: string };
			const line = (yamlErr.mark?.line ?? 0) + 1;
			const col = (yamlErr.mark?.column ?? 0) + 1;
			errors.push(
				`YAML parse error in ${file} at line ${line}, column ${col}: ${yamlErr.message ?? String(err)}`,
			);
			continue;
		}

		if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
			errors.push(`Invalid fixture format in ${file}: expected an object with 'suite' and 'tests' keys`);
			continue;
		}

		const raw = parsed as Record<string, unknown>;
		if (typeof raw['suite'] !== 'string') {
			errors.push(`Invalid fixture in ${file}: missing or non-string 'suite' key`);
			continue;
		}

		if (!Array.isArray(raw['tests'])) {
			errors.push(`Invalid fixture in ${file}: 'tests' must be an array`);
			continue;
		}

		suites.push(raw as unknown as VetoTestSuite);
	}

	return { suites, errors };
}

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

export interface RunTestsOptions {
	/** Path to fixture YAML files or directory. Defaults to ./veto/tests */
	fixturesPath?: string;
	/** Path to the policy file or directory to test against. Defaults to ./veto */
	policyPath?: string;
	/** When true, report which rule IDs have no test coverage */
	coverage?: boolean;
	/** Suppress console output */
	quiet?: boolean;
}

// ---------------------------------------------------------------------------
// Core runner
// ---------------------------------------------------------------------------

export async function runTests(options: RunTestsOptions = {}): Promise<VetoTestRunResult> {
	const fixturesPath = options.fixturesPath ?? './veto/tests';
	const policyPath = options.policyPath ?? './veto';
	const coverage = options.coverage ?? false;
	const quiet = options.quiet ?? false;

	// Load policy
	let snapshot;
	try {
		snapshot = loadPolicySnapshot(policyPath);
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		if (!quiet) console.error(colors.error(`Failed to load policy: ${msg}`));
		return { total: 0, passed: 0, failed: 0, results: [], loadErrors: [`Policy load error: ${msg}`] };
	}

	const replaySnapshot = buildReplaySnapshot(snapshot);
	const expressionCache = new Map<string, ASTNode>();

	// Load fixtures
	const { suites, errors: loadErrors } = loadSuites(fixturesPath);

	if (loadErrors.length > 0 && !quiet) {
		for (const err of loadErrors) {
			console.error(colors.error(`  Error: ${err}`));
		}
	}

	if (suites.length === 0) {
		if (!quiet) console.log(colors.warning('No test suites loaded.'));
		return { total: 0, passed: 0, failed: 0, results: [], loadErrors };
	}

	// Run tests
	const results: VetoTestResult[] = [];

	for (const suite of suites) {
		if (!quiet) console.log(`\n${colors.bold(suite.suite)}`);

		for (const testCase of suite.tests) {
			const call: ReplayCall = {
				index: 1,
				line: 1,
				tool: testCase.tool,
				arguments: testCase.arguments,
				timestamp: new Date().toISOString(),
				...(testCase.context ?? {}),
			};

			let result: VetoTestResult;
			try {
				const decision = decideReplayCall(replaySnapshot, call, [], expressionCache);
				const normalizedExpected = normalizeExpectedDecision(testCase.expect.decision);
				const actualDecision = decision.decision;
				const actualRuleId = decision.ruleId;

				const decisionMatch = actualDecision === normalizedExpected;
				const ruleIdMatch = testCase.expect.rule_id === undefined
					|| actualRuleId === testCase.expect.rule_id;
				const passed = decisionMatch && ruleIdMatch;

				result = {
					testId: testCase.id,
					suite: suite.suite,
					passed,
					expected: testCase.expect,
					actual: { decision: actualDecision, rule_id: actualRuleId },
				};

				if (!passed && !decisionMatch) {
					result.error = `expected decision '${testCase.expect.decision}' but got '${actualDecision}'`;
				} else if (!passed && !ruleIdMatch) {
					result.error = `expected rule_id '${testCase.expect.rule_id ?? ''}' but got '${actualRuleId ?? ''}'`;
				}
			} catch (err) {
				result = {
					testId: testCase.id,
					suite: suite.suite,
					passed: false,
					expected: testCase.expect,
					actual: { decision: 'error' },
					error: err instanceof Error ? err.message : String(err),
				};
			}

			results.push(result);

			if (!quiet) {
				const badge = result.passed
					? colors.success('  PASS')
					: colors.error('  FAIL');
				const desc = testCase.description ? ` — ${testCase.description}` : '';
				console.log(`${badge} ${colors.dim(testCase.id)}${desc}`);
				if (!result.passed && result.error) {
					console.log(`       ${colors.error(result.error)}`);
				}
			}
		}
	}

	const passed = results.filter((r) => r.passed).length;
	const failed = results.length - passed;

	if (!quiet) {
		console.log('');
		const summary = failed === 0
			? colors.success(`Tests: ${passed} passed`)
			: `Tests: ${colors.success(`${passed} passed`)}, ${colors.error(`${failed} failed`)}`;
		console.log(summary);
	}

	// Coverage report
	if (coverage && !quiet) {
		printCoverageReport(snapshot.rules, results);
	}

	return { total: results.length, passed, failed, results, loadErrors };
}

function printCoverageReport(
	rules: import('../rules/types.js').Rule[],
	results: VetoTestResult[],
): void {
	const testedRuleIds = new Set(
		results
			.filter((r) => r.expected.rule_id !== undefined)
			.map((r) => r.expected.rule_id as string),
	);

	const allRuleIds = rules
		.filter((r) => r.id !== undefined)
		.map((r) => r.id as string);

	const untested = allRuleIds.filter((id) => !testedRuleIds.has(id));

	console.log('');
	console.log(colors.bold('Coverage:'));
	console.log(`  Rules with tests:    ${testedRuleIds.size}`);
	console.log(`  Rules without tests: ${untested.length}`);

	if (untested.length > 0) {
		console.log('');
		console.log(colors.warning('  Untested rule IDs:'));
		for (const id of untested) {
			console.log(`    ${colors.dim(id)}`);
		}
	}
}

File: packages/sdk-python/veto/testing/runner.py
"""
Policy unit testing framework.

Reads YAML fixture files and evaluates each test case against rules
loaded from policy YAML files (no LLM, no network).
"""

from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

import yaml

from veto.testing.types import (
    TestExpect,
    VetoTestCase,
    VetoTestResult,
    VetoTestRunResult,
    VetoTestSuite,
)


def normalize_expected_decision(expected: str) -> str:
    """Normalize fixture expectations to match engine output.

    block/deny -> deny, warn/log -> allow (non-blocking).
    """
    if expected in ("block", "deny"):
        return "deny"
    if expected in ("warn", "log"):
        return "allow"
    return expected


def load_suites(fixtures_path: str) -> tuple[list[VetoTestSuite], list[str]]:
    """Load test suites from YAML files."""
    suites: list[VetoTestSuite] = []
    errors: list[str] = []
    path = Path(fixtures_path).resolve()

    if not path.exists():
        errors.append(f"Fixture path not found: {path}")
        return suites, errors

    files: list[Path] = []
    if path.is_file():
        files = [path]
    else:
        files = sorted(path.rglob("*.yaml")) + sorted(path.rglob("*.yml"))
        if not files:
            errors.append(f"No YAML fixture files found in: {path}")

    for f in files:
        try:
            content = f.read_text()
            parsed = yaml.safe_load(content)
        except Exception as e:
            errors.append(f"Error loading {f}: {e}")
            continue

        if not isinstance(parsed, dict) or "suite" not in parsed or "tests" not in parsed:
            errors.append(f"Invalid fixture format in {f}")
            continue

        test_cases: list[VetoTestCase] = []
        for tc in parsed["tests"]:
            expect_data = tc.get("expect", {})
            test_cases.append(
                VetoTestCase(
                    id=tc["id"],
                    tool=tc["tool"],
                    arguments=tc.get("arguments", {}),
                    expect=TestExpect(
                        decision=expect_data.get("decision", "allow"),
                        rule_id=expect_data.get("rule_id"),
                    ),
                    description=tc.get("description"),
                    context=tc.get("context"),
                )
            )
        suites.append(VetoTestSuite(suite=parsed["suite"], tests=test_cases))

    return suites, errors


def load_rules(policy_path: str) -> List[Dict[str, Any]]:
    """Load rules from policy YAML files."""
    rules: List[Dict[str, Any]] = []
    path = Path(policy_path).resolve()

    if path.is_file():
        files = [path]
    elif path.is_dir():
        rules_dir = path / "rules"
        if rules_dir.is_dir():
            files = sorted(rules_dir.rglob("*.yaml")) + sorted(rules_dir.rglob("*.yml"))
        else:
            files = sorted(path.rglob("*.yaml")) + sorted(path.rglob("*.yml"))
    else:
        return rules

    for f in files:
        try:
            parsed = yaml.safe_load(f.read_text())
            if isinstance(parsed, dict) and "rules" in parsed:
                rules.extend(parsed["rules"])
        except Exception:
            continue

    return rules


def evaluate_test_case(
    rules: List[Dict[str, Any]], test_case: VetoTestCase
) -> Tuple[str, Optional[str]]:
    """Evaluate a test case against rules. Returns (decision, rule_id)."""
    for rule in rules:
        if not rule.get("enabled", True):
            continue

        rule_tools = rule.get("tools", [])
        if rule_tools and test_case.tool not in rule_tools:
            continue

        conditions = rule.get("conditions", [])
        if conditions and not _evaluate_conditions(conditions, test_case):
            continue

        action = rule.get("action", "allow")
        rule_id = rule.get("id")

        if action in ("block", "deny"):
            return "deny", rule_id
        elif action == "allow":
            return "allow", rule_id
        elif action in ("warn", "log"):
            return "allow", rule_id
        elif action == "require_approval":
            return "require_approval", rule_id

    return "allow", None


def _evaluate_conditions(conditions: List[Dict[str, Any]], test_case: VetoTestCase) -> bool:
    """Simplified condition evaluator for test fixtures."""
    for cond in conditions:
        field = cond.get("field", "")
        operator = cond.get("operator", "")
        value = cond.get("value")

        if field.startswith("arguments."):
            actual = test_case.arguments.get(field[len("arguments."):])
        elif field.startswith("context."):
            actual = (test_case.context or {}).get(field[len("context."):])
        else:
            actual = test_case.arguments.get(field)
            if actual is None and test_case.context:
                actual = test_case.context.get(field)

        if operator == "equals" and actual != value:
            return False
        elif operator == "not_equals" and actual == value:
            return False
        elif operator == "contains" and (actual is None or str(value) not in str(actual)):
            return False
        elif operator == "greater_than" and (actual is None or value is None or float(actual) <= float(value)):
            return False
        elif operator == "less_than" and (actual is None or value is None or float(actual) >= float(value)):
            return False
        elif operator == "in" and actual not in (value if isinstance(value, list) else [value]):
            return False
        elif operator == "not_in" and actual in (value if isinstance(value, list) else [value]):
            return False
        elif operator == "exists" and actual is None:
            return False
        elif operator == "not_exists" and actual is not None:
            return False

    return True


def run_tests(
    fixtures_path: str = "./veto/tests",
    policy_path: str = "./veto",
    coverage: bool = False,
    quiet: bool = False,
) -> VetoTestRunResult:
    """Run policy unit tests from YAML fixtures."""
    suites, load_errors = load_suites(fixtures_path)

    if load_errors and not quiet:
        for err in load_errors:
            print(f"  Error: {err}")

    rules = load_rules(policy_path)
    results: list[VetoTestResult] = []

    for suite in suites:
        if not quiet:
            print(f"\n{suite.suite}")

        for tc in suite.tests:
            actual_decision, actual_rule_id = evaluate_test_case(rules, tc)
            normalized = normalize_expected_decision(tc.expect.decision)

            decision_match = actual_decision == normalized
            rule_id_match = tc.expect.rule_id is None or actual_rule_id == tc.expect.rule_id
            passed = decision_match and rule_id_match

            error = None
            if not decision_match:
                error = f"expected decision '{tc.expect.decision}' but got '{actual_decision}'"
            elif not rule_id_match:
                error = f"expected rule_id '{tc.expect.rule_id}' but got '{actual_rule_id}'"

            result = VetoTestResult(
                test_id=tc.id,
                suite=suite.suite,
                passed=passed,
                expected=tc.expect,
                actual_decision=actual_decision,
                actual_rule_id=actual_rule_id,
                error=error,
            )
            results.append(result)

            if not quiet:
                badge = "  PASS" if passed else "  FAIL"
                desc = f" -- {tc.description}" if tc.description else ""
                print(f"{badge} {tc.id}{desc}")
                if error:
                    print(f"       {error}")

    total_passed: int = sum(1 for r in results if r.passed)
    failed = len(results) - total_passed

    if not quiet:
        print(f"\nTests: {total_passed} passed" + (f", {failed} failed" if failed else ""))

    if coverage and not quiet:
        _print_coverage(rules, results)

    return VetoTestRunResult(total=len(results), passed=total_passed, failed=failed, results=results)


def _print_coverage(rules: List[Dict[str, Any]], results: List[VetoTestResult]) -> None:
    tested_ids = {r.expected.rule_id for r in results if r.expected.rule_id}
    all_ids = [r.get("id") for r in rules if r.get("id")]
    untested = [rid for rid in all_ids if rid not in tested_ids]

    print("\nCoverage:")
    print(f"  Rules with tests:    {len(tested_ids)}")
    print(f"  Rules without tests: {len(untested)}")

    if untested:
        print("\n  Untested rule IDs:")
        for rid in untested:
            print(f"    {rid}")
